interp/interp.go

```diff
@@ -0,0 +1,166 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package interp
+
+import (
+	"fmt"
+	"math"
+	"sort"
+
+	"gonum.org/v1/gonum/spatial/r1"
+)
+
+// Interpolator1D represents a 1-dimensional interpolator.
+// It interpolates a function over a defined closed interval range
+// (which can be infinite on both sides).
+type Interpolator1D interface {
+	// Interval returns the valid closed interval for interpolation.
+	Interval() r1.Interval
```

Comment:
Agree

Reply:
I'm not sure. Interpolation and extrapolation are two different concepts, and you often use a different algorithm for one and for the other. For example, one often uses flat extrapolation even if using a piecewise linear / cubic interpolation between the known values (to avoid the divergence of extrapolated values). If you're interpolating interest rates and the last known one is a 50year swap rate, you usually just extend it to +infinity, even if you used a completely different scheme between the known rates. On the other hand, if using piecewise linear interpolation to model a growth of something, you could extrapolate the last known slope. Baking in a fixed extrapolation scheme in every interpolator will remove the flexibility a typical user will need.

I would rather build predictors from interpolators and extrapolators then enforce that everything has to be a predictor defined over the entire real axis.

Also, extrapolation in two or more dimensions becomes more problematic.

Reply:
OK, I changed my mind - we can provide the necessary flexibility without `Interval()`. For more advanced usage, we could provide a method for concatenating two or more `Predictor` interfaces and switching between them based on the value of the argument.